src/nr_openai_observability/langchain_callback.py

```diff
@@ -0,0 +1,212 @@
+import random
+import sys
+from collections import deque
+from typing import Any, Dict, List, Optional, Union
+
+from langchain.callbacks.base import BaseCallbackHandler
+from langchain.schema import AgentAction, AgentFinish, BaseMessage, LLMResult
+from newrelic_telemetry_sdk import Span
+
+from nr_openai_observability import monitor
+
+
+class NewRelicCallbackHandler(BaseCallbackHandler):
+    def __init__(
+        self,
+        application_name: str,
+        langchain_callback_metadata: Dict[str, Any] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize callback handler."""
+        self.application_name = application_name
+
+        self.new_relic_monitor = monitor.initialization(
+            application_name=application_name,
+            parent_span_id_callback=self.parent_id_callback,  # may not be asyncio safe
+            **kwargs,
+        )
+        self.langchain_callback_metadata = langchain_callback_metadata
+        self.spans_stack = deque()
+        self.tool_invocation_counter = 0
+        self.trace_id = "%016x" % random.getrandbits(64)
+
+    def get_and_update_tool_invocation_counter(self):
+        self.tool_invocation_counter += 1
+        return self.tool_invocation_counter
+
+    def on_llm_start(
+        self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any
+    ) -> Any:
+        """Run when LLM starts running."""
+        tags = {
+            "prompt": prompts[0],
+            "model_name": kwargs.get("invocation_params", {}).get("_type", ""),
+        }
+        self.spans_stack.append(self.create_span(name="LlmCompletion", tags=tags))
+
+    def on_chat_model_start(
+        self,
+        serialized: Dict[str, Any],
+        messages: List[List[BaseMessage]],
+        **kwargs: Any,
+    ) -> Any:
+        """Run when Chat Model starts running."""
+        invocation_params = kwargs.get("invocation_params", {})
+        tags = {
+            "model": invocation_params.get("model"),
+            "model_name": invocation_params.get("model_name"),
+            "temperature": invocation_params.get("temperature"),
+            "request_timeout": invocation_params.get("request_timeout"),
+            "max_tokens": invocation_params.get("max_tokens"),
+            "stream": invocation_params.get("stream"),
+            "n": invocation_params.get("n"),
+            "temperature": invocation_params.get("temperature"),
+        }
+
+        self.spans_stack.append(self.create_span(name="LlmCompletion", tags=tags))
```

Comment:
I am wondering why this span is named the same as the span created when the LLM starts running. Since it is differentiated with a unique set of tags, should it be named accordingly? This span will also be looked for when the LLM ends or errors to be popped and recorded but shouldn't that be the span created specifically when the LLM started running?